Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm

```diff
@@ -692,21 +708,31 @@ static void dataProviderReleaseInfoCallback(void* info)
     incrementalLoaderLog(stream.release());
 #endif
 
-    WTF::Semaphore dataSemaphore { 0 };
+    Ref dataSemaphore = SemaphoreWrapper::create();
+    {
+        Locker locker { m_wasPDFThreadTerminationRequestedLock };
+        if (m_wasPDFThreadTerminationRequested)
+            return;
+
+        m_dataSemaphores.add(dataSemaphore.get());
+    }
+
     Vector<RetainPtr<CFDataRef>> dataResults(count);
 
     // FIXME: Once we support multi-range requests, make a single request for all ranges instead of <count> individual requests.
-    RunLoop::main().dispatch([protectedLoader = Ref { *this }, &dataResults, ranges, count, &dataSemaphore] {
+    RunLoop::main().dispatch([protectedLoader = Ref { *this }, &dataResults, ranges, count, dataSemaphore]() mutable {
+        Ref callbackAggregator = CallbackAggregator::create([dataSemaphore = WTFMove(dataSemaphore)] {
+            dataSemaphore->signal();
+        });
         for (size_t i = 0; i < count; ++i) {
-            protectedLoader->getResourceBytesAtPosition(ranges[i].length, ranges[i].location, [i, &dataResults, &dataSemaphore](const uint8_t* bytes, size_t bytesCount) {
+            protectedLoader->getResourceBytesAtPosition(ranges[i].length, ranges[i].location, [i, &dataResults, callbackAggregator](const uint8_t* bytes, size_t bytesCount) {
```

Comment:
🤔 not sure how `callbackAggregator` works but will it get called for each loop? if that is the case, we just `wait` one time, and it is a BinarySemaphore.

Reply:
callbackAggregator calls its lambda on destruction. Since I capture it by value in each lambda inside the loop, we will only single the semaphore ONCE, when all `getResourceBytesAtPosition()` have completed. As a result, we only need a single semaphore. This simplifies the code.